### PR TITLE
tribs changes SSCS-10587

### PR DESCRIPTION
**Pull request by nickhill111**
https://tools.hmcts.net/jira/browse/SSCS-10587

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/CcdMideventCallbackController.java [hunk 1]
```diff
@@ -97,7 +102,8 @@ public ResponseEntity<PreSubmitCallbackResponse<SscsCaseData>> ccdMidEventPrevie
 
         WriteFinalDecisionPreviewDecisionServiceBase writeFinalDecisionPreviewDecisionService = decisionNoticeService.getPreviewService(benefitType);
 
-        return ok(writeFinalDecisionPreviewDecisionService.preview(callback, DocumentType.DRAFT_DECISION_NOTICE, userAuthorisation, false));
+        return ok(writeFinalDecisionPreviewDecisionService.preview(callback,
+            SscsUtil.getWriteFinalDecisionDocumentType(caseDetails, isPostHearingEnabled), userAuthorisation, false));
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/controller/CcdMideventCallbackController.java:106]: SscsUtil.getWriteFinalDecisionDocumentType(caseDetails, isPostHearingEnabled), userAuthorisation, false)
 
 Separate var

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java [hunk 2]
```diff
@@ -162,8 +164,10 @@ private void createFinalDecisionNoticeFromPreviewDraft(PreSubmitCallbackResponse
         String now = LocalDate.now().format(DateTimeFormatter.ofPattern("dd-MM-yyyy"));
 
         final SscsDocumentTranslationStatus documentTranslationStatus = sscsCaseData.isLanguagePreferenceWelsh() ? TRANSLATION_REQUIRED : null;
-        footerService.createFooterAndAddDocToCase(documentLink, sscsCaseData, DocumentType.FINAL_DECISION_NOTICE, now,
-                null, null, documentTranslationStatus);
+        footerService.createFooterAndAddDocToCase(documentLink, sscsCaseData,
+            SscsUtil.getIssueFinalDecisionDocumentType(docLink.getDocumentFilename(), isPostHearingEnabled),
```
[the remark this answers is not in the document]
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:168]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/controller/CcdMideventCallbackController.java (hunk 1 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/controller/CcdMideventCallbackController.java:106]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToSubmitHandler.java [hunk 3]
```diff
@@ -74,7 +70,9 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
                 sscsCaseData.setPreviousState(state);
             }
 
-            previewDocumentService.writePreviewDocumentToSscsDocument(sscsCaseData, DRAFT_DECISION_NOTICE, sscsCaseData.getSscsFinalDecisionCaseData().getWriteFinalDecisionPreviewDocument());
+            previewDocumentService.writePreviewDocumentToSscsDocument(sscsCaseData,
+                SscsUtil.getWriteFinalDecisionDocumentType(caseDetails, isPostHearingEnabled),
```
[the remark this answers is not in the document]
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToSubmitHandler.java:74]: done

**Review by jarekPierchala — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java [hunk 4]
```diff
@@ -87,8 +90,9 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
         createFinalDecisionNoticeFromPreviewDraft(preSubmitCallbackResponse);
         clearTransientFields(preSubmitCallbackResponse);
 
-        if (!(State.READY_TO_LIST.equals(sscsCaseData.getState())
-            || State.WITH_DWP.equals(sscsCaseData.getState()))) {
+        if ((!(State.READY_TO_LIST.equals(sscsCaseData.getState())
+            || State.WITH_DWP.equals(sscsCaseData.getState())))
+            && !(isPostHearingsEnabled && isYes(sscsCaseData.getPostHearing().getCorrection().getCorrectionFinalDecisionInProgress()))) {
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:95]: extract variables for readability

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java [hunk 5]
```diff
@@ -102,6 +106,8 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
         if (isAdjournmentEnabled) {
             sscsCaseData.setIssueFinalDecisionDate(LocalDate.now());
         }
+
+
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:110]: remove whitespace

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToStartHandler.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        CaseDetails<SscsCaseData> caseDetails = callback.getCaseDetails();
+        SscsCaseData sscsCaseData = caseDetails.getCaseData();
+
+        SscsUtil.setCorrectionInProgress(caseDetails, isPostHearingsEnabled);
+
+        if (!isPostHearingsEnabled
+                || !isYes(sscsCaseData.getPostHearing().getCorrection().getCorrectionFinalDecisionInProgress())) {
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToStartHandler.java:44]: isNoOrNull, or if you specifically want NO, then use NO.equals(...)

File: src/main/java/uk/gov/hmcts/reform/sscs/model/docassembly/NoticeIssuedTemplateBody.java [hunk 7]
```diff
@@ -2,19 +2,20 @@
...
+import lombok.Setter;
+import lombok.experimental.SuperBuilder;
 import uk.gov.hmcts.reform.docassembly.domain.FormPayload;
 
-@Builder(toBuilder = true)
-@Value
-@JsonInclude(JsonInclude.Include.NON_NULL)
+@SuperBuilder(toBuilder = true)
+@Setter
+@Getter
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/model/docassembly/NoticeIssuedTemplateBody.java:18]: Revert this

File: src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java [hunk 8]
```diff
@@ -131,6 +132,37 @@ private static void updateTranslationStatus(SscsCaseData caseData, SscsDocumentT
...
+    public static void setCorrectionInProgress(CaseDetails<SscsCaseData> caseDetails, boolean isPostHearingsEnabled) {
+        if (isPostHearingsEnabled) {
+            Correction correction = caseDetails.getCaseData().getPostHearing().getCorrection();
+
+            if ((State.POST_HEARING.equals(caseDetails.getState())
+                    || State.DORMANT_APPEAL_STATE.equals(caseDetails.getState()))) {
+                correction.setCorrectionFinalDecisionInProgress(YesNo.YES);
+            } else {
+                correction.setCorrectionFinalDecisionInProgress(YesNo.NO);
+            }
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java:162]: Extract the if into a YesNo called correctionInProgress. Then do correction.setCorrectionFinalDecisionInProgress(correctionInProgress)

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import uk.gov.hmcts.reform.sscs.ccd.callback.Callback;
+import uk.gov.hmcts.reform.sscs.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.sscs.ccd.domain.*;
+import uk.gov.hmcts.reform.sscs.ccd.service.CcdCallbackMapService;
+import uk.gov.hmcts.reform.sscs.service.*;
+
+@RunWith(JUnitParamsRunner.class)
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:27]: Not required

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+    protected static Validator validator = Validation.byDefaultProvider()
+            .configure()
+            .messageInterpolator(new ParameterMessageInterpolator())
+            .buildValidatorFactory()
+            .getValidator();
+
+    @Before
+    public void setUp() {
+        openMocks(this);
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:51]: Remove this and use ExtendWith(MockitoExtension.class) annotation on class instead

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
+package uk.gov.hmcts.reform.sscs.ccd.presubmit.issuefinaldecision;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.*;
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:4]: Swap all Assert methods for Assertions library

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java [hunk 12]
```diff
@@ -944,18 +945,73 @@ public void givenWelsh_GeneratedDateIsAlreadySet_thenDoNotSetNewGeneratedDate()
...
+    public void givenPostHearingsIsTrue_thenShowIssuedDateOnDocument() {
+        setDescriptorFlowIndicator(isDescriptorFlowSupported() ? "yes" : "no", sscsCaseData);
+        sscsCaseData.getSscsFinalDecisionCaseData().setWriteFinalDecisionGenerateNotice("yes");
+        setHigherRateScenarioFields(sscsCaseData);
+        sscsCaseData.getSscsFinalDecisionCaseData().setWriteFinalDecisionDateOfDecision("2018-10-10");
+        sscsCaseData.getPostHearing().getCorrection().setCorrectionFinalDecisionInProgress(YesNo.YES);
+
+        service.preview(callback, DocumentType.CORRECTION_GRANTED, USER_AUTHORISATION, true, true, true);
+
+        NoticeIssuedTemplateBody payload = verifyTemplateBody(NoticeIssuedTemplateBody.ENGLISH_IMAGE, "Appellant Lastname", null, "2018-10-10", true,
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java:958]: Save this date and name as String constants - they are used throughout the test class

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java [hunk 13]
```diff
@@ -944,18 +945,73 @@ public void givenWelsh_GeneratedDateIsAlreadySet_thenDoNotSetNewGeneratedDate()
...
     protected abstract boolean isDescriptorFlowSupported();
 
     protected NoticeIssuedTemplateBody verifyTemplateBody(String image, String expectedName, String expectedAppointeeName, String dateOfDecision, boolean allowed, boolean isSetAside, boolean isDraft,
-        boolean isDescriptorFlow, boolean isGenerateFile, String templateId) {
+                                                          boolean isDescriptorFlow, boolean isGenerateFile, String templateId) {
+        return verifyTemplateBody(image, expectedName, expectedAppointeeName, dateOfDecision, allowed, isSetAside, isDraft, isDescriptorFlow, isGenerateFile, false, templateId);
+    }
+
+    protected NoticeIssuedTemplateBody verifyTemplateBody(String image, String expectedName, String expectedAppointeeName, String dateOfDecision, boolean allowed, boolean isSetAside, boolean isDraft,
+        boolean isDescriptorFlow, boolean isGenerateFile, boolean isCorrection, String templateId) {
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java:1004]: I have refactored this method in my branch to replace the boolean isDraft with a local Enum FinalDecisionType. There were a lot of usages across all the tests so much refactoring (regex replacement helped). In hindsight, an overloaded method with an extra boolean arg for isCorrection was probably the better choice. Either way, our two branches will need to be aligned.

File: src/test/java/uk/gov/hmcts/reform/sscs/util/SscsUtilTest.java [hunk 14]
```diff
@@ -15,10 +22,19 @@
 class SscsUtilTest {
     public static final String UNEXPECTED_POST_HEARING_REVIEW_TYPE_AND_ACTION = "getting the document type has an unexpected postHearingReviewType and action";
     private PostHearing postHearing;
+    private SscsCaseData caseData;
 
     @BeforeEach
     void setUp() {
-        postHearing = new PostHearing();
+        openMocks(this);
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/util/SscsUtilTest.java:29]: Use ExtendWith(MockitoExtension.class)

**Review by amoslv1 — CHANGES_REQUESTED** [review]
Some suggestions

File: src/test/java/uk/gov/hmcts/reform/sscs/util/SscsUtilTest.java (hunk 14 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/util/SscsUtilTest.java:29]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/model/docassembly/NoticeIssuedTemplateBody.java (hunk 7 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/model/docassembly/NoticeIssuedTemplateBody.java:18]: done

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java (hunk 9 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:27]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java (hunk 5 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:110]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java (hunk 4 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:95]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToStartHandler.java (hunk 6 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionAboutToStartHandler.java:44]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java (hunk 8 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java:162]: done

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java (hunk 10 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:51]: cant remove this as it complains

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java (hunk 11 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionSubmittedHandlerTest.java:4]: done

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java (hunk 13 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java:1004]: I will leave my branch as so then refactor yours when I take it over

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java (hunk 12 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceTestBase.java:958]: done

**Review by jarekPierchala — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 15]
```diff
@@ -335,7 +335,7 @@ dependencies {
     implementation group: 'com.github.hmcts', name: 'document-management-client', version: '7.0.0'
     implementation group: 'com.github.hmcts', name: 'cmc-pdf-service-client', version: '7.0.1'
 
-    implementation group: 'com.github.hmcts', name: 'sscs-common', version: '4.26.10'
+    implementation group: 'com.github.hmcts', name: 'sscs-common', version: '8d2ded7'
```
Comment (jarekPierchala) [inline build.gradle:338]: update with new common ver

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceBase.java [hunk 16]
```diff
@@ -190,21 +217,20 @@ protected abstract void setTemplateContent(DecisionNoticeOutcomeService outcomeS
...
+            if (nonNull(finalHearing) && nonNull(finalHearing.getHearingDate())) {
+                writeFinalDecisionBuilder.heldOn(LocalDate.parse(finalHearing.getHearingDate()));
             }
+        } else {
+            writeFinalDecisionBuilder.heldOn(LocalDate.now());
         }
-        writeFinalDecisionBuilder.heldOn(LocalDate.now());
-        writeFinalDecisionBuilder.heldAt("In chambers");
+
+
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceBase.java:232]: remove empty line

**Review by jarekPierchala — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java [hunk 17]
```diff
@@ -138,11 +143,86 @@ private static void updateTranslationStatus(SscsCaseData caseData, SscsDocumentT
...
+        return isPostHearingsEnabled && isYes(caseData.getPostHearing().getCorrection().getCorrectionFinalDecisionInProgress());
+    }
+      
     public static boolean isGapsCase(SscsCaseData sscsCaseData) {
         return GAPS.equals(sscsCaseData.getSchedulingAndListingFields().getHearingRoute());
     }
+
+    public static String buildWriteFinalDecisionHeldBefore(SscsCaseData caseData, @NonNull String signedInJudgeName) {
+        List<String> names = new ArrayList<>();
+
```
[the remark this answers is not in the document]
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java:185]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java [hunk 18]
```diff
@@ -138,11 +143,86 @@ private static void updateTranslationStatus(SscsCaseData caseData, SscsDocumentT
...
+            return DocumentType.CORRECTION_GRANTED;
+        }
+
+        return DocumentType.FINAL_DECISION_NOTICE;
+    }
+
+    public static void setCorrectionInProgress(CaseDetails<SscsCaseData> caseDetails, boolean isPostHearingsEnabled) {
+        if (isPostHearingsEnabled) {
+            YesNo correctionInProgress = State.POST_HEARING.equals(caseDetails.getState()) || State.DORMANT_APPEAL_STATE.equals(caseDetails.getState()) ? YES : NO;
+
```
[the remark this answers is not in the document]
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/util/SscsUtil.java:170]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceBase.java (hunk 16 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/writefinaldecision/WriteFinalDecisionPreviewDecisionServiceBase.java:232]: done

File: build.gradle (hunk 15 above)
Comment (nickhill111) [reply build.gradle]: done

**Review by DGordonV1 — APPROVED** [review]

**Review by AndreiTrushkov — APPROVED** [review]

**Review by jarekPierchala — APPROVED** [review]

**Pull request merged**